examples/pytorch/language-modeling/run_fim.py

```diff
@@ -0,0 +1,831 @@
+        )
+        model = AutoModelForCausalLM.from_pretrained(
+            model_args.model_name_or_path,
+            from_tf=bool(".ckpt" in model_args.model_name_or_path),
+            config=config,
+            cache_dir=model_args.cache_dir,
+            revision=model_args.model_revision,
+            token=model_args.token,
+            trust_remote_code=model_args.trust_remote_code,
+            torch_dtype=torch_dtype,
+            low_cpu_mem_usage=model_args.low_cpu_mem_usage,
+            attn_implementation=model_args.attn_implementation,
+        )
+
+    else:
+        model = AutoModelForCausalLM.from_config(
+            config,
+            trust_remote_code=model_args.trust_remote_code,
+            attn_implementation=model_args.attn_implementation,
+        )
+        n_params = sum({p.data_ptr(): p.numel() for p in model.parameters()}.values())
+        logger.info(f"Training new model from scratch - Total size={n_params/2**20:.2f}M params")
+
+    # Add the new FIM tokens to the tokenizer and resize model's vocab embeddings
+    special_tokens = [data_args.fim_prefix_token, data_args.fim_middle_token, data_args.fim_suffix_token]
+    if data_args.truncate_or_pad:
+        special_tokens.append(data_args.pad_token)
+
+    # Get the factor by which the embedding layer should be padded based on the device
+    pad_factor = 1
+    if torch.cuda.is_availble():
+        pad_factor = 8
+
+    elif is_torch_tpu_available():
+        pad_factor = 128
+
+    # Add the new tokens to the tokenizer
+    tokenizer.add_tokens(special_tokens)
+    # Get the pre-expansion embeddings of the model and resize the embedding layer
+    original_embeddings = model.get_input_embeddings().weight
+    model.resize_token_embeddings(len(tokenizer), pad_to_multiple_of=pad_factor)
+
+    # Sample the embeddings for the new tokens from a multivariate normal distribution
+    # We do this so that the new embeddings are close to the original embeddings and not necessarily zero
+    # More on this: https://nlp.stanford.edu/~johnhew/vocab-expansion.html
+    mean = original_embeddings.mean(dim=0)
+    n = original_embeddings.size()[0]
+    sigma = ((original_embeddings - mean).T @ (original_embeddings - mean)) / n
+    dist = torch.distributions.multivariate_normal.MultivariateNormal(
+        mean,
+        covariance_matrix=1e-5 * sigma,
+    )
+    new_token_embeddings = torch.stack(
+        tuple((dist.sample() for _ in range(len(special_tokens)))),
+        dim=0,
+    )
+
+    # Get the updated embedding layer and make a copy of it's weights
+    embeddings = model.get_input_embeddings()
+    new_embeddings = embeddings.weight.clone()
+
+    # Set the new tokens' embeddings to the newly sampled embeddings
+    new_embeddings[-len(special_tokens) :] = new_token_embeddings
+
+    # Update the model's embeddings with the new embeddings
+    embeddings.weight = torch.nn.Parameter(new_embeddings)
+    model.set_input_embeddings(embeddings)
```

Comment:
Thanks for the review @pacman100!

I had a doubt, shouldn't I be using `deepspeed.zero.GatheredParameters` for that? It will look something like this:

```python
if is_deepspeed_zero3_enabled():
    import deepspeed
    with deepspeed.zero.GatheredParameters(embeddings.weight, modifier_rank=None):
            new_embeddings = embeddings.weight.clone()

            # Set the new tokens' embeddings to the newly sampled embeddings
            new_embeddings[-len(special_tokens) :] = new_token_embeddings
        
            # Update the model's embeddings with the new embeddings
            embeddings.weight = torch.nn.Parameter(new_embeddings)
            model.set_input_embeddings(embeddings)
```

Reply:
Yes, I meant this, updated the original comment.